python/lsst/ip/isr/isrTaskLSST.py

```diff
@@ -0,0 +1,1219 @@
+        ptcDataset : `lsst.ip.isr.PhotonTransferCurveDataset`, optional
+            PTC dataset containing the gains and read noise.
+
+        Raises
+        ------
+        RuntimeError
+            Raised if either ``usePtcGains`` or ``usePtcReadNoise``
+            are ``True``, but ptcDataset is not provided.
+
+            Raised if ```doEmpiricalReadNoise`` is ``True`` but
+            ``overscanImage`` is ``None``.
+
+        See also
+        --------
+        lsst.ip.isr.isrFunctions.updateVariance
+        """
+        maskPlanes = [self.config.saturatedMaskName, self.config.suspectMaskName]
+        if self.config.usePtcGains:
+            if ptcDataset is None:
+                raise RuntimeError("No ptcDataset provided to use PTC gains.")
+            else:
+                gain = ptcDataset.gain[amp.getName()]
+                self.log.info("Using gain from Photon Transfer Curve.")
+        else:
+            gain = amp.getGain()
+
+        if math.isnan(gain):
+            gain = 1.0
+            self.log.warning("Gain set to NAN!  Updating to 1.0 to generate Poisson variance.")
+        elif gain <= 0:
+            patchedGain = 1.0
+            self.log.warning("Gain for amp %s == %g <= 0; setting to %f.",
+                             amp.getName(), gain, patchedGain)
+            gain = patchedGain
+
+        if self.config.doEmpiricalReadNoise and overscanImage is None:
+            badPixels = isrFunctions.countMaskedPixels(ampExposure.getMaskedImage(),
+                                                       [self.config.saturatedMaskName,
+                                                        self.config.suspectMaskName,
+                                                        "BAD", "NO_DATA"])
+            allPixels = ampExposure.getWidth() * ampExposure.getHeight()
+            if allPixels == badPixels:
+                # If the image is bad, do not raise.
+                self.log.info("Skipping empirical read noise for amp %s.  No good pixels.",
+                              amp.getName())
+            else:
+                raise RuntimeError("Overscan is none for EmpiricalReadNoise.")
+
+        if self.config.doEmpiricalReadNoise and overscanImage is not None:
+            stats = afwMath.StatisticsControl()
+            stats.setAndMask(overscanImage.mask.getPlaneBitMask(maskPlanes))
+            readNoise = afwMath.makeStatistics(overscanImage.getImage(),
+                                               afwMath.STDEVCLIP, stats).getValue()
+            self.log.info("Calculated empirical read noise for amp %s: %f.",
+                          amp.getName(), readNoise)
+        elif self.config.usePtcReadNoise:
+            if ptcDataset is None:
+                raise RuntimeError("No ptcDataset provided to use PTC readnoise.")
+            else:
+                readNoise = ptcDataset.noise[amp.getName()]
+            self.log.info("Using read noise from Photon Transfer Curve.")
+        else:
```

Comment:
This may also be a warning since we really only want to use certified read noises.

Reply:
I don't think this needs to be an info, it should be debug since it's the only option.  Also if it isn't supported for other read noise below why do we have the config?

Reply:
Agreed.... equally usePtcGains is also useless? I think I will get rid of them for now.

Reply:
I'm also currently on DM-37554 which will have the gain and noise defined with a function in a single point, although this is being done in `isrTask.py`. Just something to keep in mind, especially potentially going out of synch as I see this is a new file (`isrTaskLSST.py`).

Reply:
Thanks for pointing this out, as we move to the next step we'll want to keep it in sync.  But the hope with this new task is that there will be much less confusion because there will be many fewer options.  Though we probably need some way of filling a gain just for initial processing to bootstrap the PTC.  But I'd like to start with the "correct" flow and add from there, rather than holding onto legacy.